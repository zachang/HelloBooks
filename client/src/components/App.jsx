import React from 'react';
import PropTypes from 'react-proptypes';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { getOneUserAction } from '../actions/userAction';
import AdminHeader from './common/AdminHeader.jsx';
import AdminSidebar from './common/AdminSidebar.jsx';
import UserHeader from './common/UserHeader.jsx';
import UserSidebar from './common/UserSidebar.jsx';
import { decodeToken } from '../utils/helpers';

/**
 * App class declaration
 *
 * @class App
 *
 * @extends {React.Component}
 */
export class App extends React.Component {
  /**
   * @method componentWillMount
   *
   * @return {void} void
   */
  componentWillMount() {
    const userId = decodeToken(window.sessionStorage.getItem('token'));
    this.props.getOneUserAction(userId.id);
  }

  /**
   * @method componentDidMount
   *
   * @return {void} void
   */
  componentDidMount() {
    $('.button-collapse').sideNav();
    $('select').material_select();
    $('.dropdown-button').dropdown();
  }

  /**
   * Renders App component
   *
   * @return {XML} JSX
   */
  render() {
    const userRoles = decodeToken(window.sessionStorage.getItem('token'));

    return (
      <div className='row'>
        {(userRoles.isAdmin) ?
          <AdminHeader profileUsername={this.props.userState.user.username}/>
          :
          <UserHeader profileUsername={this.props.userState.user.username}/>}
        {(userRoles.isAdmin) ?
          <AdminSidebar profileImage={this.props.userState.user.userImage}/>
          :
          <UserSidebar
            profileImage={this.props.userState.user.userImage}
            userLevel={this.props.userState.user.level}
          />}
        <div className='container mainCon' style={{ marginLeft: '5%' }}>
          {this.props.children}
        </div>
      </div>
    );
  }
}

App.propTypes = {
  userState: PropTypes.object.isRequired,
  getOneUserAction: PropTypes.func.isRequired,
  children: PropTypes.object.isRequired
};
const mapStateToProps = state => ({
  userState: state.userReducer,
});
const mapDispatchToProps = dispatch =>
  bindActionCreators({ getOneUserAction }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(App);