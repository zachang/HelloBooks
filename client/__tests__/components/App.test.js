import React from 'react';
import toJson from 'enzyme-to-json';
import sinon from 'sinon';
import { shallow } from 'enzyme';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import storeConfiguration from '../../src/store/store';
import { App } from '../../src/components/App.jsx';


const categoryData = {
  categoryName: '',
};

describe('<App/>', () => {
  // create a spy function for getOneUserAction
  const getOneUserAction = sinon.spy();
  // spy on componentDidMount of App Component
  sinon.spy(App.prototype, 'componentDidMount');
  // spy on componentWillMount of App Component
  sinon.spy(App.prototype, 'componentWillMount');

  const props = {
    userState: {
      success: true,
      errors: null,
      fails: null,
      user: {
        username: 'mmmmm'
      }
    },
    getOneUserAction
  };

  const middleware = [thunk];
  const mockStore = configureMockStore(middleware);
  let store;
  let shallowComponent;
  beforeEach(() => {
    store = mockStore(storeConfiguration);
    shallowComponent = shallow(<App {...props} />);
  });

  it('renders <App /> component', () => {
    const tree = toJson(shallowComponent);
    expect(tree).toMatchSnapshot();
  });
});