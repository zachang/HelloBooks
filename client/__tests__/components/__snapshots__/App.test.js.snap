// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App/> renders <App /> component 1`] = `
<div
  className="row"
>
  <AdminHeader
    profileUsername="mmmmm"
  />
  <AdminSidebar />
  <div
    className="container mainCon"
    style={
      Object {
        "marginLeft": "5%",
      }
    }
  />
</div>
`;
